compiler/optimizer/DeadTreesElimination.cpp

```diff
@@ -508,6 +530,12 @@ void TR::DeadTreesElimination::prePerformOnBlocks()
          TR::TransformUtil::removeTree(comp(), tt);
          removed = true;
          }
+      else if (node->getOpCode().isAnchor() && visitedReadbarsUnderTreeTop.isSet(node->getFirstChild()->getGlobalIndex()))
+         {
+         TR::TreeTop *ttToRemove = rdbar2ttMap[node->getFirstChild()->getGlobalIndex()];
```

Comment:
is there a potential bug here if you have
treetop
   rdbar
CHK1
   =>rdbar
CHK2
   => rdbar

Reply:
I think it should be fine. The `rdbar2ttMap` only record nodes hanging under a `TreeTop` but not any other checks. The condition checking`if (!removed && isReadBarreirUnderTreetop(node))` when adding the rebar's treetop to the deque guarantee only the first treetop hanging the rebar will be added.

Reply:
right, the condition should check for `checks` as well.

Reply:
@andrewcraik Hey Andrew, I am reconsidering this comment https://github.com/eclipse/omr/pull/2947#discussion_r230922890. I think it shouldn't be a problem after all. Checks like `resolveCHK` or `nullCHK` has to be the first appearance of the first child so no treetop with the same child should appear before them and this pattern won't happen for those checks. For checks like `Checkcast` this tree pattern is possible and we do need the treetop as an anchor node.

Reply:
@cathyzhyi what about a SpineCHK - I don't think that requires the array child to be a first appearance so couldn't that create the above pattern?

Reply:
Had an offline discussion with Andrew and added code to make the code more conservative. Do not remove the treetop if there is any other tree in between the first treetop and the anchoring treetop. something like:
treetop
=>rdbar
ANYTREE
=>rdbar
anchor
=>rdbar

Reply:
@vijaysun-omr I think this more conservative version of just handling the treetop and anchor pattern should be sufficient - thoughts?

Reply:
I think this is safe since it's unclear what ANYTREE assumptions may be wrt anchoring earlier.